bite-size-notes/week-6-d-Linear MPC application VS LQR

```diff
@@ -0,0 +1,214 @@
+- **Prediction Model:**
+This model predicts future outputs of the system based on current and past information and the proposed control inputs. It usually encompasses the dynamics of the vehicle or process being controlled.
+- **Control Inputs:**
+These are the actions or commands generated by the optimizer that are sent to the vehicle or process. These inputs are designed to drive the system towards the reference objectives.
+- **Vehicle:**
+This represents the physical system or process being controlled by the MPC. In this context, it's labeled as a "vehicle," which could be literal (such as a car or drone) or figurative (such as any system that 'moves' or changes state).
+- **Measured Outputs:**
+These are the actual outputs from the vehicle or process, which are fed back into the MPC controller. This feedback loop allows the MPC to update its predictions and optimize future control inputs, adjusting for any discrepancies between the predicted and actual outputs.
+
+
+## MPC VS LQR application
+As we have learned about LQR in the previous section, now we want to see the strength and weakness of this MPC model compared to LQR algorithm :
+#### **Linear Quadratic Regulator (LQR)**
+LQR is a feedback control strategy designed for linear systems and aims to minimize a quadratic cost function that represents the error (deviation from the desired state) and the control effort. The solution to the LQR problem involves solving an algebraic Riccati equation to find an optimal gain matrix \( K \). This gain matrix is used to compute the control input as \( u = -Kx \), where \( x \) is the state vector of the system.
+
+#### **Model Predictive Control (MPC)**
+MPC solves an optimization problem at each step in the control process. It uses a model of the system to predict future states over a finite horizon based on current states and a sequence of future control inputs. The objective is to minimize a cost function typically similar to LQR but over a finite horizon and subject to constraints on states and controls. After solving the optimization, only the first control action is implemented, and the process is repeated at the next step.
+
+#### Advantages of Using MPC Over LQR
+- **Better at handling constraints on inputs and states,** ensuring compliance with operational limits.
+- **Adapts in real-time to changes and disturbances,** improving system responsiveness and robustness.
+- **Utilizes predictive capabilities to anticipate future events,** optimizing control actions proactively.
+
+#### Disadvantages of Using MPC Over LQR
+- **Requires significant computational resources,** which can be a constraint in real-time applications.
+- **Implementation complexity is higher** due to the need for solving optimization problems frequently.
+- **Relies heavily on model accuracy,** with performance degrading if the predictive model is not precise.
+
+### MPC VS LQR Implementation Example
```

Comment:
MPC vs. LQR Implementation Example:
The code snippets lack detailed explanations of variables and functions, making it challenging for readers unfamiliar with programming or control theory to follow along.
The simulation results could be analyzed and discussed in more detail to provide deeper insights into the performance differences between LMPC and LQR.

Reply:
Thanks, I have made a brief explanation and summary about the code. I was worried on explaining everything from the beginning since the code is very long. I hope my brief explanation helps the reader to understand